Adds GS Distribution Test and Tracer

nils-wisiol opened the pull request.

nils-wisiol commented on victims/GS/tracer.py:
```diff
@@ -0,0 +1,31 @@
+
+    def trigger(self, seed: str, poly_length: int = 12) -> Tuple:
+        seed = self.random_seed(f'GS {seed}', 16) 
+        samples_length = 2
```
Let's do one sample at a time, so we can understand the individual traces easier

nils-wisiol commented on victims/GS/test_GS.py:
```diff
@@ -1,15 +1,22 @@
+        results = victim.receive_array('q', little=True)[1]
+
+        # Check if the statistic of result are correct
+        mu, sigma = norm.fit(results)  # noqa
```
We need to assert somehow that the distribution approximately fits a Gaussian distribution

nils-wisiol commented on victims/GS/test_GS.py:
```diff
@@ -1,15 +1,22 @@
+    samples_length = [512]
+    for i in range(1):
+        # Create a random key
+        k = list(default_rng(256).integers(low=0, high=2**8 - 1, size=16))
```
The `256` here is a bit confusing: it's the seed for the PRNG, it has nothing to do with the output size. Any number would do, why choose this one?

nils-wisiol commented:
We probably need to select the tracing time more carefully: currently, the samples takes longer to generate two samples than we can trace. (Max trace length is about 24k.)

Even after so many samples, the process apparently didn't finish:

![image](https://user-images.githubusercontent.com/1026940/87765629-550e6b00-c818-11ea-9cb7-014508b1e454.png)

nils-wisiol commented on victims/GS/main.c:
```diff
@@ -1,96 +1,125 @@
-     status = gaussian_sampler(result,result_length);
-     trigger_low();
+    status = randomness_set_seed(key);
+    trigger_high();
```
Move this to after AES

Soundes-M commented:
> We probably need to select the tracing time more carefully: currently, the samples takes longer to generate two samples than we can trace. (Max trace length is about 24k.)
> 
> Even after so many samples, the process apparently didn't finish:
> 
> ![image](https://user-images.githubusercontent.com/1026940/87765629-550e6b00-c818-11ea-9cb7-014508b1e454.png)

I moved the trigger after the randomness_generator. I also tried to "optimize" the code and use smaller variables without affecting the accuracy of the results. (see the ode I reduced from uint64_t to uint8_t). That did not change the traces' length.  
I had an idea to trace only the rejection condition (see the last commit) instead of (sampling+rejection). However I am not sure if we can predict the sample by doing that.

nils-wisiol commented:
>  (see the last commit)

I think you didn't add any new commits?

Soundes-M commented:
> > (see the last commit)
> 
> I think you didn't add any new commits?
on TestGS branch. DId you move the code to TestGS2?

nils-wisiol commented:
Yeah [I took the code from TestGS,](https://github.com/nils-wisiol/m4sc/pull/17#issuecomment-659949109) rebased it on top of master, cleaned it up, removed some merge errors and put it in TestGS2.

nils-wisiol commented on victims/GS/test_GS.py:
```diff
@@ -1,22 +1,33 @@
-        # Create a random key
-        k = list(default_rng(256).integers(low=0, high=2**8 - 1, size=16))
-
+    samples_length = [1024]
```
I believe this won't work on chipwhisperer

nils-wisiol commented on victims/GS/main.c:
```diff
@@ -38,19 +38,56 @@ static uint8_t recv_key(int size, uint16_t *data) {
+     {  
+         
+        
+	do
```
adjust formatting in the following lines :)

nils-wisiol commented on victims/GS/test_GS.py:
```diff
@@ -1,15 +1,33 @@
+        victim.receive_return_code()
+
+        # Create a random key
+        k = list(default_rng().integers(low=0, high=2 ** 16 - 1, size=16))
```
add a seed argument to `default_rng` to obtain reproducible results
